applications/StructuralMechanicsApplication/tests/test_truss_adjoint_element_3d2n.py

```diff
@@ -0,0 +1,288 @@
+    mp_old.Nodes[node_id].GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_X,0))
+    node.SetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Y,
+    mp_old.Nodes[node_id].GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Y,0))
+    node.SetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Z,
+    mp_old.Nodes[node_id].GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Z,0))
+
+def shape_perturbation_correction_factor(node_1, node_2):
+    dx = node_1.X - node_2.X
+    dy = node_1.Y - node_2.Y
+    dz = node_1.Z - node_2.Z
+    l = math.sqrt(dx*dx + dy*dy + dz*dz)
+    return l    
+  
+def zero_vector(size):
+    v = KratosMultiphysics.Vector(size)
+    for i in range(size):
+        v[i] = 0.0
+    return v
+
+def create_property_perturbed_elements(mp, delta, new_element_name):
+    dim = 3
+    model_part = KratosMultiphysics.ModelPart("Property_Perturbed_Elements")
+    model_part.ProcessInfo.SetValue(KratosMultiphysics.DOMAIN_SIZE,dim)
+    add_variables(model_part)
+    model_part.CreateNewNode(1, mp.Nodes[1].X, mp.Nodes[1].Y, mp.Nodes[1].Z)
+    model_part.CreateNewNode(2, mp.Nodes[2].X, mp.Nodes[2].Y, mp.Nodes[2].Z)
+    apply_material_properties(model_part,dim)
+
+    A_initial = mp.GetProperties()[1][StructuralMechanicsApplication.CROSS_AREA]
+    model_part.GetProperties()[1].SetValue(StructuralMechanicsApplication.CROSS_AREA, A_initial + delta )
+    prop = model_part.GetProperties()[1]
+
+    model_part.CreateNewElement(new_element_name, 1, [1, 2], prop)
+
+    for i in range(2):
+        copy_solution_step_data_of_node(model_part.Nodes[i+1], mp, i+1, 0)
+
+    for element in model_part.Elements:
+        element.Initialize()
+
+    return model_part
+
+def create_shape_perturbed_elements(mp, delta, new_element_name):
+    dim=3
+    model_part = KratosMultiphysics.ModelPart("Shape_Perturbed_Elements")
+    add_variables(model_part)
+
+    x1 = mp.Nodes[1].X
+    y1 = mp.Nodes[1].Y
+    z1 = mp.Nodes[1].Z
+    x2 = mp.Nodes[2].X
+    y2 = mp.Nodes[2].Y
+    z2 = mp.Nodes[2].Z
+    model_part.CreateNewNode(1, x1, y1, z1)
+    model_part.CreateNewNode(2, x1+delta, y1, z1)
+    model_part.CreateNewNode(3, x1, y1+delta, z1)
+    model_part.CreateNewNode(4, x1, y1, z1+delta)
+    model_part.CreateNewNode(5, x2, y2, z2)
+    model_part.CreateNewNode(6, x2+delta, y2, z2)
+    model_part.CreateNewNode(7, x2, y2+delta, z2)
+    model_part.CreateNewNode(8, x2, y2, z2+delta)
+
+    apply_material_properties(model_part,dim)
+    prop = model_part.GetProperties()[1]
+
+    model_part.CreateNewElement(new_element_name, 1, [2, 5], prop)
+    model_part.CreateNewElement(new_element_name, 2, [3, 5], prop)
+    model_part.CreateNewElement(new_element_name, 3, [4, 5], prop)
+    model_part.CreateNewElement(new_element_name, 4, [1, 6], prop)
+    model_part.CreateNewElement(new_element_name, 5, [1, 7], prop)
+    model_part.CreateNewElement(new_element_name, 6, [1, 8], prop)
+
+    index = 1
+    for i in range(2):
```

Comment:
this is hard to understand => `zip` & `enumerate`

Reply:
modified